Option for building both Python 2 and Python 3 modules

MrAnno opened the pull request:
This PR adds the possibility to build syslog-ng with both the Python 2 and Python 3 support enabled:

```
./configure --enable-python --with-python=2,3
```

This can be used to package both the Python 2 and Python 3 modules, so the user can choose between versions. As the Python 2 EOL is announced, this has very little use, but it could make the Python 2 -> Python 3 transition easier for the user.

syslog-ng compiled with this option will start the Python 2 driver if `python()` or `python2()` is specified in the configuration. `python3()` is for Python 3 plugins.

`python()` currently means Python 2, but it can be deprecated and removed in the future.

In case `--with-python=2,3` is _not_ used, the old behavior remains: `python()` launches the driver syslog-ng was configured with.

The PR also makes persist names, statistic counters, and library names "versioned" to avoid confusion and different kind of collisions.
Non-versioned persist names are automatically upgraded at startup (python destination).

If you upgrade syslog-ng from source (without using a package manager),  execute `make uninstall` before installing the new version, as the old (non-versioned) Python module may be left in the module directory.

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

Kokan commented on configure.ac:
```diff
@@ -1406,7 +1410,15 @@ PKG_CHECK_MODULES(libsystemd, libsystemd >= ${LIBSYSTEMD_MIN_VERSION},
+
+if test "x$enable_python2_and_3" = "xyes"; then
+  PKG_CHECK_MODULES(PYTHON3, python3,, AC_MSG_ERROR([Could not find Python 3 development libraries]))
+  PKG_CHECK_MODULES(PYTHON2, python2,, AC_MSG_ERROR([Could not find Python 2 development libraries]))
```
The auto detection checks for `python3`, `python2` and `python` in that order, as sometimes there is only `python` (which is actually `python2`) and does not exists `python2` thus this fails.
I would consider checking `python` if the `python2` is not found.
Well I question remains that does `python` always means `python2` or not ?! (sorry I do not know the answer)

MrAnno replied:
Sure, I'll add the `python` check.

Kokan commented on modules/python/Makefile.am:
```diff
@@ -69,6 +70,28 @@ modules_python_libmod_python_la_DEPENDENCIES = $(MODULE_DEPS_LIBS)
 
+else
+
+module_LTLIBRARIES += modules/python/libmod-python2.la
```
[optional]
I would prefer to keep the `libmod-python.la` name, and define as `libmod-python2.la` (almost arbitrary choose)
Because some cases it is a possibility to have both `libmod-python{,2,3}.la`, where at least two of them defines the same module name as well. (optional because I only could come up the case when developers do switching between branches options)

MrAnno replied:
Good idea.

kira-syslogng commented:
Build SUCCESS

MrAnno commented:
CMake is on the way, I'll update the PR one last time.

kira-syslogng commented:
Build SUCCESS

furiel commented on modules/python/python-dest.c:
```diff
@@ -119,9 +119,9 @@ python_dd_format_stats_instance(LogThreadedDestDriver *d)
 
   if (d->super.super.super.persist_name)
-    g_snprintf(persist_name, sizeof(persist_name), "python,%s", d->super.super.super.persist_name);
+    g_snprintf(persist_name, sizeof(persist_name), PYTHON_MODULE_PERSIST_KEY ",%s", d->super.super.super.persist_name);
```
Do we need to change persist? Simply having `python` in any case looks neat to me.

MrAnno replied:
It's hard to tell. In my opinion, persist keys should be as unique as possible. We fortunately have key collision detection in OSE, but a Python 2 module and Python 3 module is different in my mind. Upgrading a source/destination/parser from v2 to v3 or backwards has to be a deliberate action, I'd force the user to use `persist-tool` or `persist-name()` to make the transition if they need the old persist entry.

furiel replied:
One last argument that might worth considering is inconsistency. If I see a persist name of `python`, I cannot tell if it comes from
- `--with-python=2 --disable-python2_and_3`
- `--with-python=3 --disable-python2_and_3`
- `--enable-python2_and_3` but python2 is used.

Currently `PYTHON_MODULE_PERSIST_KEY` persist does not provide uniqueness among python versions. To provide uniqueness, you need to eliminate `python`, and provide `python2` and `python3` consistently.

MrAnno replied:
Yes, this is for compatibility. I'd like to remove the support of Python 3 with the unversioned python keyword as soon as possible.

MrAnno commented on modules/python/Makefile.am:
```diff
@@ -69,6 +70,28 @@ modules_python_libmod_python_la_DEPENDENCIES = $(MODULE_DEPS_LIBS)
+
+module_LTLIBRARIES += modules/python/libmod-python2.la
+modules_python_libmod_python2_la_CFLAGS = $(python_common_cflags) $(PYTHON2_CFLAGS)
+modules_python_libmod_python2_la_CPPFLAGS = -DSYSLOG_NG_ENABLE_PYTHONv2=1 $(python_common_cppflags)
```
They can't. When `--enable-python2-and-3` is used, nor `SYSLOG_NG_ENABLE_PYTHONv2` neither `SYSLOG_NG_ENABLE_PYTHONv3` is set, it has to be set here in our Makefile.

That was my idea to avoid additional complex ifdef branching in our codebase.

MrAnno commented on modules/python/python-source.c:
```diff
@@ -561,7 +561,7 @@ python_sd_new(GlobalConfig *cfg)
   self->super.format_stats_instance = python_sd_format_stats_instance;
   self->super.worker_options.super.stats_level = STATS_LEVEL0;
-  self->super.worker_options.super.stats_source = stats_register_type("python");
+  self->super.worker_options.super.stats_source = stats_register_type(PYTHON_MODULE_PERSIST_KEY);
```
Sure, I don't see why not.

MrAnno commented on modules/python/compat/compat-python-v3.c:
```diff
@@ -23,6 +23,30 @@
+void
+_set_python_home(const gchar *python_home)
+{
+  Py_SetPythonHome(Py_DecodeLocale(python_home, NULL));
```
Oh, nice catch. However, we don't have such destructor hook in syslog-ng. Should I use a gcc destructor?

SetPythonHome has this requirement: `... whose contents will not change for the duration of the program’s execution.`

furiel replied:
A little unfortunate, but I understand, and you can leave as it is from my side. This is a one time leak (not leaks during reload).

MrAnno replied:
`app_startup()` and `app_shutdown()` hooks for modules could help a lot in such cases.

MrAnno commented on configure.ac:
```diff
@@ -1859,6 +1871,8 @@ AC_DEFINE_UNQUOTED(SYSTEMD_JOURNAL_MODE, `journald_mode`, [Systemd-journal suppo
 AC_DEFINE_UNQUOTED(HAVE_INOTIFY, `enable_value $ac_cv_func_inotify_init`, [Have inotify])
 AC_DEFINE_UNQUOTED(ENABLE_PYTHONv2, `enable_value $enable_python2`, [Enable Python 2 support])
 AC_DEFINE_UNQUOTED(ENABLE_PYTHONv3, `enable_value $enable_python3`, [Enable Python 3 support])
+AC_DEFINE_UNQUOTED(ENABLE_PYTHONv2_AND_v3, `enable_value $enable_python2_and_3`, [Enable both Python 2 and Python 3 support])
```
I chose this method to avoid additional ifdefing in our code base (it's already complex enough there), so I can compile the same library twice setting the appropriate macro in Makefile (`SYSLOG_NG_ENABLE_PYTHONv{2,3}`).

MrAnno commented on modules/python/python-config.c:
```diff
@@ -63,11 +63,11 @@ python_config_new(void)
     {
       pc = python_config_new();
-      g_hash_table_insert(cfg->module_config, g_strdup(PYTHON_MODULE_CONFIG_KEY), pc);
+      g_hash_table_insert(cfg->module_config, g_strdup(PYTHON_MODULE_PERSIST_KEY), pc);
```
I think this is not weird in case of module-config. CfgPersist, "real" persist and and module-config should use the same persist key in my opinion.

bazsi commented:
I think, if at all possible, we should use explicit versions and upgrade
earlier unversioned persist values to explicit python2/python3 names
probably depending on the platform in question (e.g. how the syslog-ng
binary on that platform is assumed to have been compiled, I know this is
not a definite answer, but would handle the majority of the cases). This is
an interesting article on this topic:

https://lwn.net/Articles/780737/

On the upgrade path, we should probably be allowing python2 -> python3
upgrades, but not the other way around.

On Thu, Aug 22, 2019 at 8:02 AM furiel <notifications@github.com> wrote:

> *@furiel* commented on this pull request.
> ------------------------------
>
> In modules/python/python-dest.c
> <https://github.com/balabit/syslog-ng/pull/2845#discussion_r316507922>:
>
> > @@ -119,9 +119,9 @@ python_dd_format_stats_instance(LogThreadedDestDriver *d)
>    static gchar persist_name[1024];
>
>    if (d->super.super.super.persist_name)
> -    g_snprintf(persist_name, sizeof(persist_name), "python,%s", d->super.super.super.persist_name);
> +    g_snprintf(persist_name, sizeof(persist_name), PYTHON_MODULE_PERSIST_KEY ",%s", d->super.super.super.persist_name);
>
> One last argument that might worth considering is inconsistency. If I see
> a persist name of python, I cannot tell if it comes from
>
>    - --with-python=2 --disable-python2_and_3
>    - --with-python=3 --disable-python2_and_3
>    - --enable-python2_and_3 but python2 is used.
>
> Currently PYTHON_MODULE_PERSIST_KEY persist does not provide uniqueness
> among python versions. To provide uniqueness, you need to eliminate python,
> and provide python2 and python3 consistently.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2845?email_source=notifications&email_token=AAFOK5VW62JGSIECFAS3XP3QFYTYTA5CNFSM4IG6SJDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCKREQA#discussion_r316507922>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFOK5QOPD34H2DVS3TJ3PDQFYTYTANCNFSM4IG6SJDA>
> .
>


-- 
Bazsi

MrAnno commented:
@bazsi Thanks for the article. I agree, we should use versioned names, I'd also like to deprecate the config keyword `python()` at least for Python 3 in favor of `python3()`.

The `python2 -> python3` persist upgrade is a good idea, but I'd do it in a separate PR, if it's not a problem.

bazsi commented:
It's fine in a separate pr

On Thu, Aug 22, 2019, 13:13 László Várady <notifications@github.com> wrote:

> @bazsi <https://github.com/bazsi> Thanks for the article. I agree, we
> should use versioned names, I'd also like to deprecate the config keyword
> python() at least for Python 3 in favor of python3().
>
> The python2 -> python3 persist upgrade is a good idea, but I'd do it in a
> separate PR, if it's not a problem.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2845?email_source=notifications&email_token=AAFOK5RUDHWXCW6OF5ZEIPLQFZYGLA5CNFSM4IG6SJDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD44XWWQ#issuecomment-523860826>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFOK5UB5SDZFCZZGWT7J3LQFZYGLANCNFSM4IG6SJDA>
> .
>

bazsi commented on modules/python/Makefile.am:
```diff
@@ -1,75 +1,97 @@
+modules_python_libmod_python_la_SOURCES = $(python_sources)
+modules_python_libmod_python_la_LDFLAGS = $(MODULE_LDFLAGS) $(PYTHON2_LDADD)
+modules_python_libmod_python_la_DEPENDENCIES = $(MODULE_DEPS_LIBS)
+
```
With this change I would try to make the Python version explicit in our python module naming. Since we have module auto-loading, the changing of the module name shouldn't matter from a user's perspective.

e.g. we should never generate an unversioned python module, it should be libmod-python2 and libmod-python3. At the very least this should resolve potential file naming conflicts.

I would also remove ENABLE_PYTHONv2_AND_v3 conditional and only use ENABLE_PYTHON2 and ENABLE_PYTHON3, and maybe couple the same with configure command line changes (e.g. separate --enable-python2 and --enable-python3 options), with both autodetected by default.

Then the only point for potential conflict would be our use of the "python" keyword, e.g. which version should be picked when the user uses the python keyword, which has been our only keyword in the past. I have the following ideas:

1) use a configure option and specify versions explicitly for this case  and use the version we are compiled with on the prominent platforms (for instance CzanikP's RHEL6 packages are compiled with Python3), 
2) pick the one that the distro uses when launching the unversioned python (can be detected at configure time)
3) the same but also honour distro configuration at runtime (IIRC Fedora allows the default to be changed)

Maybe some of these can be combined without getting insane and getting a lot of complexity and least amount of breakage down the line.

bazsi commented:
This is great stuff!

furiel commented:
@MrAnno could you resolve the conflict?

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

MrAnno commented:
@bazsi Sorry for the huge delay. 

> With this change I would try to make the Python version explicit in our python module naming. Since we have module auto-loading, the changing of the module name shouldn't matter from a user's perspective.

Done. I've added an uninstall hook for the old `libmod-python.so`, just to avoid different collision issues when syslog-ng is installed from source (without using a package manager).

> I would also remove ENABLE_PYTHONv2_AND_v3 conditional and only use ENABLE_PYTHON2 and ENABLE_PYTHON3, and maybe couple the same with configure command line changes (e.g. separate --enable-python2 and --enable-python3 options), with both autodetected by default.

I replaced this with `--with-python=2,3`. We already have `--enable-python` and `--with-python=x`, I did not want to complicate or break things.

Persist, stats, and everything now use versioned Python names, except for the config:
`python()` means Python 2, unless syslog-ng is compiled only with Python 3 support. `python2()` and `python3()` are also available.

>1) use a configure option and specify versions explicitly for this case  and use the version we are compiled with on the prominent platforms (for instance CzanikP's RHEL6 packages are compiled with Python3), 
>2) pick the one that the distro uses when launching the unversioned python (can be detected at configure time)
>3) the same but also honour distro configuration at runtime (IIRC Fedora allows the default to be changed)

I've only implemented 1). The change is already unpleasant, I tried not to complicate things, but I also did not want to break anything (other than the stats names). The non-versioned -> versioned persist name update is automatically done at startup (Python destination).


Python 2 is EOL, so we aren't sure it's worth the complexity that this PR would add to the Python module. I asked for the opinion of you and the syslog-ng Debian package maintainers about this topic in gcsideal/syslog-ng-debian/pull/5.

czanik commented:
Two info:
* RHEL 8 will support Python 2 for another four years, until the end of 2024
* I switched all of my official distro packages and unofficial syslog-ng repos exclusively to Python 3 almost two years ago and nobody asked me for Python 2 support (this covers Fedora/RHEL, openSUSE/SLES and FreeBSD, so roughly 80%+ of our users)
